util/movement-types/src/algs/grouping_heuristic/mod.rs

```diff
@@ -0,0 +1,458 @@
+pub mod splitting;
+pub mod chunking;
+pub mod apply;
+pub mod drop_success;
+pub mod binpacking;
+pub mod skip;
+
+use std::fmt::Debug;
+
+/// A failure type for a single member of the heuristically formed group.
+pub enum ElementalFailure<T> {
+    /// An instrumental failure is intended to be be passed on in future iterations.
+    Instrumental(T),
+    /// A terminal failure is intended to be dropped or pause the execution altogether.
+    Terminal(T)
+}
+
+impl <T> Debug for ElementalFailure<T>
+where T: Debug {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ElementalFailure::Instrumental(t) => write!(f, "Instrumental({:?})", t),
+            ElementalFailure::Terminal(t) => write!(f, "Terminal({:?})", t)
+        }
+    }
+}
```

Comment:
This is a poorly re-implemented `#[derive(Debug)]` that does not use `f.debug_tuple(...)` helpers. Just use the derive?